### PR TITLE
8305056: Avoid unaligned access in emit_intX methods if it's unsupported

### PR DESCRIPTION
**Pull request by VladimirKempik**
Please review this change which attempts to eliminate unaligned memory stores generated by emit_int16/32/64 methods on some platforms.

Primary aim is risc-v platform. But I had to change some code in ppc/arm32/x86 to prevent possible perf degradation.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305056](https://bugs.openjdk.org/browse/JDK-8305056): Avoid unaligned access in emit_intX methods if it's unsupported

### Reviewers
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13227/head:pull/13227` \
`$ git checkout pull/13227`

Update a local copy of the PR: \
`$ git checkout pull/13227` \
`$ git pull https://git.openjdk.org/jdk.git pull/13227/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13227`

View PR using the GUI difftool: \
`$ git pr show -t 13227`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13227.diff">https://git.openjdk.org/jdk/pull/13227.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13227#issuecomment-1488541574)

### CODE REVIEW

**Comment by jcking** [comment]
`UseUnalignedAccesses` is a flag, no? Doesn't this now require checking to see if its aligned or the flag is true? So for aligned accesses the speed is the same, for unaligned it is now slower as it has to look at the flag first, which is likely somewhere else in memory, forcing cache lines to be flushed.

**Comment by VladimirKempik** [comment]
> Sorry if I'm slow, but I do not understand this patch, and the JBS issue text does not help much. Why would the existing test for alignment not be sufficient to prevent unaligned access?

Hello

Currently emit_intXX perform unaligned access without any check for existing flags regarding the status of unaligned access.
This especially bad(performance wise) on platforms where misalignment access is emulated (e.g. some risc-v boards).

The idea of the patch is to use put_native_uX instead of direct pointer deref.

But then I need to adjust some existing put_native_uX code to respect UseUnalignedAccess flags.


As for speed of new put_native after adding flag check - that flag is kind of constant during the lifetime of jvm and should be easy for any branch predictor, resulting in a pretty low overhead.

**Comment by tstuefe** [comment]
> > Sorry if I'm slow, but I do not understand this patch, and the JBS issue text does not help much. Why would the existing test for alignment not be sufficient to prevent unaligned access?
> 
> Hello
> 
> Currently emit_intXX perform unaligned access without any check for existing flags regarding the status of unaligned access. This especially bad(performance wise) on platforms where misalignment access is emulated (e.g. some risc-v boards).
> 
> The idea of the patch is to use put_native_uX instead of direct pointer deref.
> 
> But then I need to adjust some existing put_native_uX code to respect UseUnalignedAccess flags.
> 
> As for speed of new put_native after adding flag check - that flag is kind of constant during the lifetime of jvm and should be easy for any branch predictor, resulting in a pretty low overhead.

@VladimirKempik Thank you for the clarification!

**Comment by VladimirKempik** [comment]
> It would probably be more efficient if you just use `memcpy` and let the compiler figure out the best method to do memory accesses.

Yeah, I think it would be better to just use memcpy in emit_intX and to not modify any put_native_uY code at all

File: src/hotspot/share/asm/codeBuffer.hpp [hunk 1]
```diff
@@ -248,11 +255,11 @@ class CodeSection {
...
 
-  void emit_int64( uint64_t x)  { *((uint64_t*) end()) = x; set_end(end() + sizeof(uint64_t)); }
+  void emit_int64( uint64_t x)  { put_native(end(), x); set_end(end() + sizeof(uint64_t)); }
 
-  void emit_float( jfloat  x)  { *((jfloat*)  end()) = x; set_end(end() + sizeof(jfloat)); }
-  void emit_double(jdouble x)  { *((jdouble*) end()) = x; set_end(end() + sizeof(jdouble)); }
-  void emit_address(address x) { *((address*) end()) = x; set_end(end() + sizeof(address)); }
+  void emit_float( jfloat  x)  { put_native(end(), jint_cast(x)); set_end(end() + sizeof(jfloat)); }
+  void emit_double(jdouble x)  { put_native(end(), julong_cast(x)); set_end(end() + sizeof(jdouble)); }
+  void emit_address(address x) { put_native(end(), p2i(x)); set_end(end() + sizeof(address)); }
```
Comment (RealFYang) [inline src/hotspot/share/asm/codeBuffer.hpp:262]: You might want to remove the explicit casts of 'x'.

File: src/hotspot/share/asm/codeBuffer.hpp [hunk 2]
```diff
@@ -218,7 +225,7 @@ class CodeSection {
     set_end(curr);
   }
 
-  void emit_int16(uint16_t x) { *((uint16_t*) end()) = x; set_end(end() + sizeof(uint16_t)); }
+  void emit_int16(uint16_t x) { put_native(end(), x); set_end(end() + sizeof(uint16_t)); }
```
Comment (theRealAph) [inline src/hotspot/share/asm/codeBuffer.hpp:228]:
```suggestion
  void emit_int16(uint16_t x) { put_native(end(), x); set_end(end() + sizeof x; }
```

File: src/hotspot/share/asm/codeBuffer.hpp [hunk 3]
```diff
@@ -35,6 +35,13 @@
 #include "utilities/resizeableResourceHash.hpp"
 #include "utilities/macros.hpp"
 
+template <typename T>
+static inline void put_native(address p, T x) {
+    assert(p != nullptr, "null pointer");
+
+    memcpy((void*)p, &x, sizeof(T));
```
Comment (theRealAph) [inline src/hotspot/share/asm/codeBuffer.hpp:42]:
```suggestion
    memcpy((void*)p, &x, sizeof x);
```

File: src/hotspot/share/asm/codeBuffer.hpp (hunk 1 above)
Comment (theRealAph) [inline src/hotspot/share/asm/codeBuffer.hpp:262]:
```suggestion
  template <typename T>
  void emit_native(T x) { put_native(end(), x); set_end(end() + sizeof x); }
  void emit_float( jfloat  x)  { emit_native(x); }
  void emit_double(jdouble x)  { emit_native(x); }
  void emit_address(address x) { emit_native(x); }
```

**Comment by VladimirKempik** [comment]
the numbers on misaligned stores ( trp_sam event) before and after the patch on risc-v:

before:

 Performance counter stats for './jdk/bin/java -version':

            170054      trp_lam                                                     
             34171      trp_sam                                                     

       5.120392220 seconds time elapsed

       5.865599000 seconds user
       0.593732000 seconds sys

after:

            169598      trp_lam                                                     
             13562      trp_sam                                                     

       5.374909022 seconds time elapsed

       6.057279000 seconds user
       0.524302000 seconds sys

**Comment by VladimirKempik** [comment]
Windows failures are infra issue ( visual studio can't be installed) and unrelated.

**Review by theRealAph — COMMENTED** [review]
This looks good. I'm sure GCC and LLVM will be fine, but it might be worth checking that Windows doesn't generate awful code for `put_native()`.If you don't have a Microsoft box, ping George.Adams@microsoft.com and Bruno Borges <Bruno.Borges@microsoft.com> ?

**Comment by VladimirKempik** [comment]
yet to check win64 jvm.dll.
so far godbolt shows me this for msvc 19 -O2

void put_native<unsigned short>(void *,unsigned short) PROC ; put_native<unsigned short>, COMDAT
  mov WORD PTR [rcx], dx
  ret 0
void put_native<unsigned short>(void *,unsigned short) ENDP ; put_native<unsigned short>

**Review by theRealAph — APPROVED** [review]
Looks good to me.

**Comment by VladimirKempik** [comment]
/integrate

I will keep on eye on emit_intX code in win64  for some time to make sure it's correctly inlined.

**Pull request closed without merging**